docs/framework-design.md

```diff
@@ -0,0 +1,145 @@
+
+### Criteria for inclusion
+
+In general, a file should be in the framework repository if it is expected to be present with near-identical content in two or more consuming repositories. Some files have a significant proportion of shared content and branch-specific content; such files should be split into a shared part and a non-shared part.
+
+For example:
+
+* `test_suite_*.function` contains a lot of code that is specific to each branch of each consuming repository. Even when the same test function exists in all maintained branches, there are often minor differences such as a deprecated alternative that only exists in older branches, differences in compile-time dependencies, etc. Thus we do not expect to share these files. Common code can go into separate `.c` files, historically under `tests/src`, that are in the framework repository.
+* `test_suite_*.data` contains many test cases that exist in all maintained branches. However the exact expression of these test cases are often different, for example due to compile-time dependencies. Furthermore the set of test cases is often different, for example due to cryptographic mechanisms that are added or removed. Thus we do not expect to share these files. Where there is a lot of commonality, the test cases can be generated from a script located in the framework repository, with code in the generation script to handle the differing parts.
+
+## CI architecture
+
+Tasks:
+
+* CI in consuming repositories must support Git submodules. Other than that, keep the CI as it is now. In particular, the CI in consuming repositories does not need to consider anything but the commit that the framework submodule points to.
+* CI in the framework repository should run a subset of the CI of all consuming repositories, to warn about unintended breakage. This way, most of the time, updating the framework submodule in a consuming repository to the tip of the `main` branch should work. Gatekeepers can bypass this check if the incompatibility is deliberate.
+* When merging a pull request to in an official branch in a consuming repository (`development`, LTS branches), check that the commit.
+
+TODO: once this is set up, detail the processes here.
+
+## How to make a change
+
+### Change in a consuming repository requiring a new framework feature
+
+If a change in a consuming repository requires a new feature in the framework, you need to make both a pull request in the framework repository and a pull request in the framework repository.
+
+1. Make a pull request (PR) in the framework repository.
+2. Upload the framework branch to the framework repository itself (not a fork). This is necessary for the commit to be available on the CI of the consuming repositories (and also for it to be conveniently available to reviewers).
+   Open question: can we make the CI work with a fork, and make using forks convenient enough for reviewers, so that people don't need to upload the branch to the main repository?
```

Comment:
Follow-up on my previous question: `refs/pull/*/head` are commits in the main repository thus by using them the CI would work only with commits in the main repository, right?

Reply:
They are commits in the main repository, but not fetched by default. On the CI we use Git via a Jenkins plugin and I don't know offhand how flexible it is. I intend to leave this as an open question in this pull request, to be settled by y'all.